`StringIndexOutOfBoundsException` thrown when trying to perform `Substring.Replacement`

This looks a little like https://github.com/scala-steward-org/scala-steward/pull/3016 maybe, but that fix was merged for Scala Steward 0.24.0, and this run is [running with Scala Steward v0.26.0](https://github.com/guardian/scala-steward-public-repos/actions/runs/6800738143/job/18489978384#step:5:39).

The full exception in the [logs](https://github.com/guardian/scala-steward-public-repos/actions/runs/6800738143/job/18489978384#step:5:2479) is:

```
  2023-11-08 16:01:30,316 ERROR Steward guardian/play-secret-rotation failed
  java.lang.Throwable: failed to edit /home/runner/scala-steward/workspace/repos/guardian/play-secret-rotation/build.sbt
  	at org.scalasteward.core.io.FileAlg$$anonfun$editFile$4.applyOrElse(FileAlg.scala:69)
  	at org.scalasteward.core.io.FileAlg$$anonfun$editFile$4.applyOrElse(FileAlg.scala:69)
  	at scala.PartialFunction$AndThen.applyOrElse(PartialFunction.scala:288)
  	at cats.ApplicativeError.$anonfun$recoverWith$1(ApplicativeError.scala:172)
  	at flatMap @ org.scalasteward.core.io.FileAlg.editFile(FileAlg.scala:68)
  	at adaptErr$extension @ org.scalasteward.core.io.process$.$anonfun$slurp$3(process.scala:68)
  	at flatTraverse @ org.scalasteward.core.update.artifact.ArtifactMigrationsLoader.loadAll(ArtifactMigrationsLoader.scala:40)
  	at flatTraverse @ org.scalasteward.core.update.artifact.ArtifactMigrationsLoader.loadAll(ArtifactMigrationsLoader.scala:40)
  	at flatTraverse @ org.scalasteward.core.update.artifact.ArtifactMigrationsLoader.loadAll(ArtifactMigrationsLoader.scala:40)
  	at flatMap @ org.scalasteward.core.edit.EditAlg.$anonfun$applyUpdateReplacements$3(EditAlg.scala:108)
  	at map @ org.scalasteward.core.edit.EditAlg.applyUpdateReplacements(EditAlg.scala:106)
  	at flatMap @ org.scalasteward.core.edit.EditAlg.applyUpdateReplacements(EditAlg.scala:106)
  	at flatMap @ org.scalasteward.core.edit.EditAlg.$anonfun$applyUpdate$7(EditAlg.scala:64)
  	at flatMap @ org.scalasteward.core.edit.EditAlg.$anonfun$applyUpdate$5(EditAlg.scala:61)
  	at flatMap @ org.scalasteward.core.edit.EditAlg.$anonfun$applyUpdate$4(EditAlg.scala:58)
  	at map @ org.scalasteward.core.edit.EditAlg.$anonfun$applyUpdate$1(EditAlg.scala:56)
  	at flatMap @ org.scalasteward.core.edit.EditAlg.$anonfun$applyUpdate$1(EditAlg.scala:56)
  	at flatMap @ fs2.Compiler$Target.flatMap(Compiler.scala:163)
  Caused by: java.lang.StringIndexOutOfBoundsException: begin 2326, end 2322, length 3903
  	at java.base/java.lang.String.checkBoundsBeginEnd(String.java:3319)
  	at java.base/java.lang.String.substring(String.java:1874)
  	at org.scalasteward.core.edit.update.data.Substring$Replacement$.$anonfun$applyAll$3(Substring.scala:48)
  	at scala.collection.immutable.List.foreach(List.scala:333)
  	at org.scalasteward.core.edit.update.data.Substring$Replacement$.applyAll(Substring.scala:47)
  	at org.scalasteward.core.edit.EditAlg.$anonfun$applyUpdateReplacements$5(EditAlg.scala:109)
  	at org.scalasteward.core.io.FileAlg.$anonfun$editFile$3(FileAlg.scala:68)
  	at scala.Option.fold(Option.scala:263)
  	at org.scalasteward.core.io.FileAlg.$anonfun$editFile$1(FileAlg.scala:68)
  	... 14 common frames omitted
  2023-11-08 16:01:30,318 INFO  ──────────── Total time: Steward guardian/play-secret-rotation: 33s 573ms ────────────
```

Note that exception says the substring has `begin 2326, end 2322` - which means that the requested end was 4 characters _before_ the requested start.

This seems to be failing consistently since I merged https://github.com/guardian/play-secret-rotation/pull/411 at 12:02pm today:

![image](https://github.com/scala-steward-org/scala-steward/assets/52038/bd0b7485-2f7d-400d-88a9-9e45d00e8abf)

This is the current build.sbt file in that project:

https://github.com/guardian/play-secret-rotation/blob/07068e712ab5aa8a41cc6354c34b1eadd58dd206/build.sbt

I'll probably have to exclude this particular project from our Scala Steward run until this is fixed, otherwise we get some [bad side-effects from consistent failure](https://github.com/guardian/scala-steward-public-repos/issues/60).

--- a/modules/core/src/main/scala/org/scalasteward/core/edit/update/ModulePositionScanner.scala
+++ b/modules/core/src/main/scala/org/scalasteward/core/edit/update/ModulePositionScanner.scala
@@ -36,16 +36,16 @@ object ModulePositionScanner {
       fileData: FileData
   ): Iterator[ModulePosition] =
     sbtModuleIdRegex(dependency).findAllIn(fileData.content).matchData.map { m =>
-      val groupId = Substring.Position.fromMatch(fileData.path, m, dependency.groupId.value)
-      val artifactId = Substring.Position.fromMatch(fileData.path, m, dependency.artifactId.name)
-      val version = Substring.Position.fromMatch(fileData.path, m, m.group(1))
+      val groupId = Substring.Position(fileData.path, m.start(1), dependency.groupId.value)
+      val artifactId = Substring.Position(fileData.path, m.start(2), dependency.artifactId.name)
+      val version = Substring.Position(fileData.path, m.start(3), m.group(3))
       ModulePosition(groupId, artifactId, version)
     }
 
   private def sbtModuleIdRegex(dependency: Dependency): Regex = {
     val g = Regex.quote(dependency.groupId.value)
     val a = Regex.quote(dependency.artifactId.name)
-    raw""""$g"\s*%+\s*"$a"\s*%+\s*(.*)""".r
+    raw""""($g)"\s*%+\s*"($a)"\s*%+\s*(.*)""".r
   }
 
   private def findMillDependency(
@@ -53,16 +53,16 @@ object ModulePositionScanner {
       fileData: FileData
   ): Iterator[ModulePosition] =
     millDependencyRegex(dependency).findAllIn(fileData.content).matchData.map { m =>
-      val groupId = Substring.Position.fromMatch(fileData.path, m, dependency.groupId.value)
-      val artifactId = Substring.Position.fromMatch(fileData.path, m, dependency.artifactId.name)
-      val version = Substring.Position.fromMatch(fileData.path, m, m.group(1))
+      val groupId = Substring.Position(fileData.path, m.start(1), dependency.groupId.value)
+      val artifactId = Substring.Position(fileData.path, m.start(2), dependency.artifactId.name)
+      val version = Substring.Position(fileData.path, m.start(3), m.group(3))
       ModulePosition(groupId, artifactId, version)
     }
 
   private def millDependencyRegex(dependency: Dependency): Regex = {
     val g = Regex.quote(dependency.groupId.value)
     val a = Regex.quote(dependency.artifactId.name)
-    raw"""["`]$g:+$a:+(.*)["`;]""".r
+    raw"""["`]($g):+($a):+(.*)["`;]""".r
   }
 
   private def findMavenDependency(
@@ -70,15 +70,15 @@ object ModulePositionScanner {
       fileData: FileData
   ): Iterator[ModulePosition] =
     mavenDependencyRegex(dependency).findAllIn(fileData.content).matchData.map { m =>
-      val groupId = Substring.Position.fromMatch(fileData.path, m, dependency.groupId.value)
-      val artifactId = Substring.Position.fromMatch(fileData.path, m, dependency.artifactId.name)
-      val version = Substring.Position.fromMatch(fileData.path, m, m.group(2))
+      val groupId = Substring.Position(fileData.path, m.start(1), dependency.groupId.value)
+      val artifactId = Substring.Position(fileData.path, m.start(2), dependency.artifactId.name)
+      val version = Substring.Position(fileData.path, m.start(4), m.group(4))
       ModulePosition(groupId, artifactId, version)
     }
 
   private def mavenDependencyRegex(dependency: Dependency): Regex = {
     val g = Regex.quote(dependency.groupId.value)
     val a = Regex.quote(dependency.artifactId.name)
-    raw"""<groupId>$g</groupId>\s*<artifactId>$a(|_[^<]+)</artifactId>\s*<version>(.*)</version>""".r
+    raw"""<groupId>($g)</groupId>\s*<artifactId>($a)(|_[^<]+)</artifactId>\s*<version>(.*)</version>""".r
   }
 }

--- a/modules/core/src/test/scala/org/scalasteward/core/edit/RewriteTest.scala
+++ b/modules/core/src/test/scala/org/scalasteward/core/edit/RewriteTest.scala
@@ -890,55 +890,60 @@ class RewriteTest extends FunSuite {
     runApplyUpdate(update, original, expected)
   }
 
-  test("issue-2877: sbt using same version in a val and a literal using a Seq addition") {
+  // https://github.com/scala-steward-org/scala-steward/issues/2877
+  test("sbt using same version in a val and a literal using a Seq addition") {
     val update = ("org.scalatest".g % Nel.of(
       "scalatest".a,
       "scalactic".a
     ) % "3.2.13" %> "3.2.14").group
     val original = Map(
       "build.sbt" ->
-        """
-          |val ScalaTestVersion = "3.2.13"
+        """val ScalaTestVersion = "3.2.13"
           |libraryDependencies ++= Seq(
           |  "org.scalatest" %% "scalatest" % ScalaTestVersion,
           |  "org.scalatest" %% "scalactic" % "3.2.13"
-          |)
-          |""".stripMargin
+          |)""".stripMargin
     )
     val expected = Map(
       "build.sbt" ->
-        """
-          |val ScalaTestVersion = "3.2.14"
+        """val ScalaTestVersion = "3.2.14"
           |libraryDependencies ++= Seq(
           |  "org.scalatest" %% "scalatest" % ScalaTestVersion,
           |  "org.scalatest" %% "scalactic" % "3.2.14"
-          |)
-          |""".stripMargin
+          |)""".stripMargin
     )
     runApplyUpdate(update, original, expected)
   }
 
-  test("issue-2877: sbt using same version in a val and a literal using individual additions") {
+  // https://github.com/scala-steward-org/scala-steward/issues/2877
+  test("sbt using same version in a val and a literal using individual additions") {
     val update = ("org.scalatest".g % Nel.of(
       "scalatest".a,
       "scalactic".a
     ) % "3.2.13" %> "3.2.14").group
     val original = Map(
       "build.sbt" ->
-        """
-          |val ScalaTestVersion = "3.2.13"
+        """val ScalaTestVersion = "3.2.13"
           |libraryDependencies += "org.scalatest" %% "scalatest" % ScalaTestVersion
           |libraryDependencies += "org.scalatest" %% "scalactic" % "3.2.13"
           |""".stripMargin
     )
     val expected = Map(
       "build.sbt" ->
-        """
-          |val ScalaTestVersion = "3.2.14"
+        """val ScalaTestVersion = "3.2.14"
           |libraryDependencies += "org.scalatest" %% "scalatest" % ScalaTestVersion
           |libraryDependencies += "org.scalatest" %% "scalactic" % "3.2.14"
           |""".stripMargin
     )
+    runApplyUpdate(update, original, expected)
+  }
+
+  // https://github.com/scala-steward-org/scala-steward/issues/3206
+  test("sbt module where the artifactId is also part of the groupId") {
+    val update = ("com.typesafe.play".g % "play".a % "2.9.0" %> "3.0.0").single
+      .copy(newerGroupId = Some("org.playframework".g), newerArtifactId = Some("play"))
+    val original = Map("build.sbt" -> """ "com.typesafe.play" %% "play" % "2.9.0" """)
+    val expected = Map("build.sbt" -> """ "org.playframework" %% "play" % "3.0.0" """)
     runApplyUpdate(update, original, expected)
   }
 

--- a/modules/core/src/test/scala/org/scalasteward/core/edit/update/ModulePositionScannerTest.scala
+++ b/modules/core/src/test/scala/org/scalasteward/core/edit/update/ModulePositionScannerTest.scala
@@ -33,4 +33,19 @@ class ModulePositionScannerTest extends FunSuite {
     )
     assertEquals(obtained, expected)
   }
+
+  test("sbt module where the artifactId is also part of the groupId") {
+    val d = "com.typesafe.play".g % "play".a % "2.9.0"
+    val fd =
+      FileData("build.sbt", s""""com.typesafe.play" %% "play" % "2.9.0"""")
+    val obtained = ModulePositionScanner.findPositions(d, fd)
+    val expected = List(
+      ModulePosition(
+        Substring.Position(fd.path, 1, d.groupId.value),
+        Substring.Position(fd.path, 24, d.artifactId.name),
+        Substring.Position(fd.path, 32, s"\"${d.version}\"")
+      )
+    )
+    assertEquals(obtained, expected)
+  }
 }

--- a/modules/core/src/test/scala/org/scalasteward/core/edit/update/ModulePositionScannerTest.scala
+++ b/modules/core/src/test/scala/org/scalasteward/core/edit/update/ModulePositionScannerTest.scala
@@ -36,8 +36,7 @@ class ModulePositionScannerTest extends FunSuite {
 
   test("sbt module where the artifactId is also part of the groupId") {
     val d = "com.typesafe.play".g % "play".a % "2.9.0"
-    val fd =
-      FileData("build.sbt", s""""com.typesafe.play" %% "play" % "2.9.0"""")
+    val fd = FileData("build.sbt", s""""com.typesafe.play" %% "play" % "2.9.0"""")
     val obtained = ModulePositionScanner.findPositions(d, fd)
     val expected = List(
       ModulePosition(

